hail/python/hail/methods/impex.py

```diff
@@ -1293,12 +1293,68 @@ def import_gen(path,
     -------
     :class:`.MatrixTable`
     """
-    path = wrap_to_list(path)
+    
+    gen_table = import_lines(path, min_partitions)
+    sample_table = import_lines(sample_file)
     rg = reference_genome.name if reference_genome else None
+    contig_recoding = contig_recoding
     if contig_recoding is None:
-        contig_recoding = {}
-    return MatrixTable(ir.MatrixRead(ir.MatrixGENReader(
-        path, sample_file, chromosome, min_partitions, tolerance, rg, contig_recoding, skip_invalid_loci)))
+        contig_recoding = hl.empty_dict(hl.tstr, hl.tstr)
+    else:
+        contig_recoding = hl.dict(contig_recoding)
+
+    gen_table = gen_table.transmute(data=gen_table.text.split(' '))
+
+    if chromosome is None:
+        index = 5
+        contig_holder = gen_table.data[0]
+    else:
+        index = 4
+        contig_holder = chromosome
+
+    contig_holder = contig_recoding.get(contig_holder, contig_holder)
+
+    if rg is None:
+        locus = hl.struct(contig=contig_holder, position=hl.int(gen_table.data[index - 2]))
+    else:
+        if skip_invalid_loci:
+            locus = hl.if_else(hl.is_valid_locus(contig_holder, hl.int(gen_table.data[index - 2]), rg),
+                               hl.locus(contig_holder, hl.int(gen_table.data[index - 2]), rg),
+                               hl.missing(hl.tlocus(rg)))
+        else:
+            locus = hl.locus(contig_holder, hl.int(gen_table.data[index - 2]), rg)
+
+    alleles = hl.array([hl.str(gen_table.data[index - 1]), hl.str(gen_table.data[index])])
+    rsid = gen_table.data[index - 3]
+    varid = gen_table.data[index - 4]
+    gen_table = gen_table.annotate(locus=locus, alleles=alleles, rsid=rsid, varid=varid)
+    gen_table = gen_table.annotate(entries=gen_table.data[index + 1:].map(lambda x: hl.float64(x))
+                                   .grouped(3).map(lambda x: hl.struct(GP=x)))
+
+    sample_table_count = sample_table.count() - 2
```

Comment:
Maybe just add a comment here explaining this 2. (I know why it's there, but for the next person.